Implement third party auth ID mapping API

xcompass opened the pull request:
**Description:**
This implements a new API that can be used to retrieve a list of mappings between third party auth remote IDs and edX usernames. For use cases and more details, please see [proposal + discussion](https://docs.google.com/document/d/1Uifq_uDRfPUzba7mikWo_F_S89ynDwkQ6D0-SXhXl9g/edit?usp=sharing).

* This PR is based on @bradenmacdonald 's PR https://github.com/edx/edx-platform/pull/9672. Large part of the code has been reviewed there.
* The current authentication for this API is using either EDX_API_KEY or OAuth2 Authorization.

**Background:**
When an edX course is used in an on-campus environment and SAML, LTI or other SSO is enabled to allow students to authenticate to edX using their on-campus credentials, normally there will be an identifier sent from an on-campus IdP. This identifier is stored in edX and linked to the edX account. Later when students return from an on-campus IdP, edX is able to authenticate the student using their edX account. 

Because students may create edX accounts not using their real identity, the instructors will have difficulty to identify the edX students in their courses and won't be able to upload grades from edX back to on-campus gradebook system as the downloaded grades from edX only contain edX IDs, which are unknown to on-campus system.

This new API can be used for mapping between the edX user ID and the ID provided by identity provider (IdP). This API will give information that will allow instructors or course support staff to figure out the identity of the on-campus students who are using the edX course. It will also allow grade information coming from edX to be appropriately associated to campus students for uploading to an on-campus learning management system (LMS) or student information system (SIS). 

This API is suppose to be consumed by a on-campus middleware, which combines the information from on-campus system. The middleware would talk to campus system to get information about course and enrolment. Then use those information to control instructor access. So the instructor will be able to get his/her edX student on-campus identity, uploading grades back to their LMS or SIS, making groups based on on-campus activities, etc.

This API also helps with troubleshooting with the issue where students put in wrong email addresses and weren't able to activate their accounts. Thus they weren't able to login EdX through Shibboleth because of the inactive accounts. With this API, on-campus course support staff or instructor can provide EdX support the student real EdX username instead of the ID that Shibboleth passed over to EdX, thus avoid the EdX support manually querying the tables to figure out the mapping between Shibboleth ID and EdX username.

**JIRA:** N/A

**Component Affected:** LMS (Third Party Auth)

**Users Affected:** Currently users with EDX_API_KEY (EdX internal) or OAuth2 client authorization (external use)

**Timeframe:** Mid Oct.

**Reviewers:** @bradenmacdonald and TBD

**Resources Needed:** None

**Database Table Changes:**

* This PR adds a new table called third_party_auth_apipermissions. So a `paver update_db` is required after merging.

**Test Instructions:**

* Enable one or more third party auth providers (such as 'Dummy') and link them to user's accounts.
* Setup authentication and authorization:
  * For using EDX_API_KEY: Edit ~/lms.auth.json and set EDX_API_KEY to some value, e.g. api-key. And restart LMS.
  * For using OAuth2:
    * Setup an OAuth2 Client. In Django Admin, go to OAuth2 -> Client -> Add Client. 
      * User:  it is good idea to create a dedicated user for this API
      * URL, Redirect URL: http://localhost// (not being used in this API)
      * Client Type: Confidential
      * Other fields: no specific requirements
    * Create an authorization code. As EdX/django-auth2-provider doesn't support client_credential grant type, we need to create the authorization code manually. Go to OAuth2 -> Grants -> Add grant. Remember the code for later use.
    * Give the new client permission to this API. Go to Third_Party_Auth -> Provider API Permissions -> Add Provider API Permission
      * Select the new client and the provider created in the previous step. This will give the new client access to query that provider mapping.
* Test this API with commands like below, where {provider_id} should be replaced with the actual ID. If it is SAML provider, provider ID should be "saml-PROVIDER_SLUG", e.g., saml-ubc. For OAuth2 providers, it will be the "oa2-BACKEND_NAME", e.g., oa2-google
  * Using EDX_API_KEY
    ```bash
    curl --header "X-edx-API-key: api-key" http://localhost:8000/api/third_party_auth/v0/providers/{provider_id}/users 
    
    curl --header "X-edx-API-key: api-key" http://localhost:8000/api/third_party_auth/v0/providers/{provider_id}/users?username=USERNAME1,USERNAME2

    curl --header "X-edx-API-key: api-key" http://localhost:8000/api/third_party_auth/v0/providers/{provider_id}/users?username=USERNAME1&username=USERNAME2
    
    curl --header "X-edx-API-key: api-key" http://localhost:8000/api/third_party_auth/v0/providers/{provider_id}/users?remote_id=REMOTE_ID1,REMOTE_ID2

    curl --header "X-edx-API-key: api-key" http://localhost:8000/api/third_party_auth/v0/providers/{provider_id}/users?remote_id=REMOTE_ID1&remote_idREMOTE_ID2
    ```

  * Using OAuth2
    * Using the authorization code created in the previous step to get the access token
    ```bash
    curl --data "grant_type=authorization_code&code=CODE&client_id=CLIENT_ID&client_secret=CLIENT_SECRET" http://localhost:8000/oauth2/access_token
    Return: {"access_token": "2983fbbc6c41947a2807946678d594b06b403dbf", "token_type": "Bearer", "expires_in": 31535999, "refresh_token": "909c2b0926d642283f762f7437c1a9cd12844b07", "scope": ""}
    ```
    * Using access token to issue request to API:
    ```bash
    curl --header "Authorization: Bearer ACCESS_TOKEN" http://localhost:8000//api/third_party_auth/v0/providers/{provider_id}/users

    curl --header "Authorization: Bearer ACCESS_TOKEN" http://localhost:8000//api/third_party_auth/v0/providers/{provider_id}/users?username=USERNAME1,USERNAME2

    curl --header "Authorization: Bearer ACCESS_TOKEN" http://localhost:8000//api/third_party_auth/v0/providers/{provider_id}/users?username=USERNAME1&username=USERNAME2

    curl --header "Authorization: Bearer ACCESS_TOKEN" http://localhost:8000//api/third_party_auth/v0/providers/{provider_id}/users?remote_id=REMOTE_ID1,REMOTE_ID2

    curl --header "Authorization: Bearer ACCESS_TOKEN" http://localhost:8000//api/third_party_auth/v0/providers/{provider_id}/users?remote_id=REMOTE_ID1&remote_id=REMOTE_ID2
    ```
* API should return a json object including the mappings. E.g.:
```json
{
  "page": 1,
  "page_size": 200,
  "count": 8,
  "results": [
    {"username": "USERNAME1", "remote_id": "REMOTE_ID1"},
    {"username": "USERNAME2", "remote_id": "REMOTE_ID2"},
  ]
}
```

edx-webhook commented:
Thanks for the pull request, @xcompass! I've created [OSPR-820](https://openedx.atlassian.net/browse/OSPR-820) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

xcompass commented:
ping @antoviaque

sarina commented:
Hi @xcompass - since this is a work in progress, could you let me know what resources, if any, you need from edX at this time?

Could you also flesh out your PR description with more detail, especially regarding timeline (when do you need this feature by, and please be very explicit), how to test the feature, etc? Check out our detailed guidelines here: http://edx-developer-guide.readthedocs.org/en/latest/process/cover-letter.html - making specific note of our brand new guidelines on how to contribute to the documentation for your feature.

antoviaque commented:
@xcompass Thanks for the ping! We can schedule one of the two reviews on our side - likely over the coming week.

xcompass commented:
@sarina I've updated the PR description. Let me know if I missed anything. Thanks.

xcompass commented:
@sarina updated based on your comments. Is it easier for you to see the updates if I squash the comments?

sarina commented:
Separate commits are easier, I think. :+1: from me on test coverage and code quality. However, Braden and a platform team member will need to give thumbs up as well.

bradenmacdonald commented:
@xcompass FYI, I will continue reviewing this for you once you add a model or mechanism for defining permissions that control access to this API (which is the missing piece needed for OAuth support, as we discussed).

xcompass commented:
@bradenmacdonald I've done the authorization part. Could you take a look at it when you get some time? Thanks!

xcompass commented:
@nasthagiri I have some trouble to get authorization_code from command line (curl). What I'm doing right now is manually creating the code in Django admin. But once I use the code to get the token, the code is expired. (which is expected behaviour). Do you have any tip to get the authorization code from command line? Thanks.

nasthagiri commented:
@xcompass To support server-to-server auth, we need to use the `client_credentials` grant (not the `authorization_code` grant - sorry for misspeaking earlier).  This is described in http://tools.ietf.org/html/rfc6749#section-4.4 and http://alexbilbie.com/2013/02/a-guide-to-oauth-2-grants/.
Unfortunately, it doesn't seem that this grant_type is supported in the version of the django-oauth2-provider that we use.  Do you have any luck in finding a django implementation of this that we can incorporate into our oath provider?

By the way, the `EDX_API_KEY` is something that we want to deprecate.  It provides an all-or-nothing access to our server APIs and something we want to replace in favor of OAuth2+JWTs in the long run.

xcompass commented:
@nasthagiri Thanks. I thought I missed something. Yep, I figured current django-oauth2-provider doesn't support client_credentials.

It seems django-oauth2-toolkit supports client_credentials grant. 
https://github.com/evonove/django-oauth-toolkit/blob/8b8e638d2fd0c5dd4a5595573542a476cd4a6c9d/oauth2_provider/oauth2_validators.py#L22. Also the current version of Django Rest Framework is recommanding it over django-oauth2-provider: http://www.django-rest-framework.org/api-guide/authentication/#django-oauth-toolkit

I never used it before, so I'm not sure how much work involved. Are you guys interested in migrating to django-oauth2-toolkit?

nasthagiri commented:
> It seems django-oauth2-toolkit supports client_credentials grant. 
> Are you guys interested in migrating to django-oauth2-toolkit?
Migrating to `django-oauth2-toolkit` may be a large effort, perhaps outside the scope of this PR.
However, I do strongly prefer using the `client_credentials` grant rather than a shared all-or-nothing API_KEY in the long run.  We can also consider updating the `django-oauth2-provider` library to add support for client_credentials. 

One thought is to address using the `client_credentials` grant in a separate PR.  We will also want to deprecate the use of the EDX_API_KEY throughout our platform.

bradenmacdonald commented:
@nasthagiri @xcompass Is it possible for us to somehow use the `authorization_code` grant in the meantime to authorize UBC's server to pull data from this API?

nasthagiri commented:
@bradenmacdonald Yes, it's "possible".  But it's not how the `authorization_code` grant type was intended to be used.   It's possible by having UBC's server make the one-time call to get the access_token, store the access_token for all future calls, and then refresh the access_token whenever it expires.

bradenmacdonald commented:
@nasthagiri I understand. I'm just trying to figure out how we can get this merged on time so that UBC has something they can use within their timeframe. Do you think we should use `authorization_code` for now (and perhaps you could create some stories for edX to migrate to django-oauth2-toolkit in the future), or do you see another path forward?

FYI @xcompass @nasthagiri @sarina I'm going to be away next week. So if we don't get this to the point where I can give +1 today, Eugeny will likely take over from me as a reviewer.
I'm also wondering, who will be the one to merge this PR when it is ready?

xcompass commented:
> Yes, it's "possible". But it's not how the authorization_code grant type was intended to be used. It's possible by having UBC's server make the one-time call to get the access_token, store the access_token for all future calls, and then refresh the access_token whenever it expires.

@nasthagiri @bradenmacdonald This is what I'm doing when developing this authorization.

nasthagiri commented:
Pan, that sounds good for now.  We will then update edx-platform to support the client_credentials grant type.  Thanks.

Sent from my iPhone

> On Sep 26, 2015, at 5:07 PM, Pan Luo <notifications@github.com> wrote:
> 
> Yes, it's "possible". But it's not how the authorization_code grant type was intended to be used. It's possible by having UBC's server make the one-time call to get the access_token, store the access_token for all future calls, and then refresh the access_token whenever it expires.
> 
> @nasthagiri @bradenmacdonald This is what I'm doing when developing this authorization.
> 
> —
> Reply to this email directly or view it on GitHub.
>

xcompass commented:
@bradenmacdonald Updated based on your comments.

bradenmacdonald commented on common/djangoapps/third_party_auth/provider.py:
```diff
@@ -52,6 +52,11 @@ def get(cls, provider_id):
     @classmethod
+    def get_all_provider_ids(cls):
+        """ Gets a list of all enabled provider ids """
+        return [provider.provider_id for provider in cls._enabled_providers()]
```
Please remove this method (`get_all_provider_ids`) and its test as well, if it's unused.

bradenmacdonald commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -89,3 +98,123 @@ def get(self, request, username):
+            query = query | Q(user__username__in=usernames)
+        if remote_ids_str:
+            remote_ids = [self.provider.get_social_auth_uid(remote_id) for remote_id in remote_ids_str.split(',')]
+            query = query | Q(uid__in=remote_ids)
```
Should these not be `query = query & Q(...)`? If I specify `?usernames=bob&remote_ids=123`, I would expect that it only returns a result if `bob` exists and his `remote_id == 123`. But the way this is coded now, I could actually get two or more results.

xcompass replied:
No, it should return two items. {username: bob, remote_id: 456}, {username: john, remote_id=123}. So we can do mixed mapping in one call. If you already know both username and remote_id, you don't need to map them, do you? Though you may want to use it to verify the existence of the user, but you can use single value to do so.

bradenmacdonald commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -89,3 +98,123 @@ def get(self, request, username):
+            # multi-IdP backend, we only retrieve the ones that are for current IdP
+            uid = self.provider.get_social_auth_uid('uid')
+            if uid is not 'uid':
+                query = query | Q(uid__startswith=uid[:-3])
```
This filter should always be applied, not just when no `username_str` or `remote_ids_str` is present (and use `&` not `|`).

I have two SAML providers, `testshiba` and `testshibB`, and the `myself` user has connected his account to both of those. His entries in the UserSocialAuths table like this:
```
mysql> SELECT username, provider, uid FROM social_auth_usersocialauth, auth_user WHERE auth_user.id=user_id and user_id=16;
+-----------+----------+------------------+
| username  | provider | uid              |
+-----------+----------+------------------+
| myselfEdX | tpa-saml | testshibB:myself |
| myselfEdX | tpa-saml | testshiba:myself |
+-----------+----------+------------------+
```

When I run the command `curl --header "Authorization: Bearer 675..." http://localhost:8000/api/third_party_auth/v0/providers/saml-testshiba/users?usernames=myselfEdX` 

I get this traceback:

```
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 109, in get_response
    response = callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/compat.py", line 137, in view
    return self.dispatch(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 77, in wrapped_view
    return view_func(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py", line 400, in dispatch
    response = self.handle_exception(exc)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py", line 397, in dispatch
    response = handler(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/generics.py", line 425, in get
    return self.list(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/mixins.py", line 98, in list
    return Response(serializer.data)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/serializers.py", line 575, in data
    self._data = self.to_native(obj)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/serializers.py", line 351, in to_native
    value = field.field_to_native(obj, field_name)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/serializers.py", line 410, in field_to_native
    return [self.to_native(item) for item in value.all()]
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/serializers.py", line 351, in to_native
    value = field.field_to_native(obj, field_name)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/fields.py", line 1035, in field_to_native
    value = getattr(self.parent, self.method_name)(obj)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/third_party_auth/api/serializers.py", line 23, in get_remote_id
    return self.provider.get_remote_id_from_social_auth(social_user)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/third_party_auth/models.py", line 319, in get_remote_id_from_social_auth
    assert self.match_social_auth(social_auth)
AssertionError
```

The reason is that my URL specifies the `testshiba` provider only, but the result set contains two rows, and the serializer asks `testshiba` to serialize both. But `testshiba` asserts when asked to serialize the UID `testshibB:myself`

bradenmacdonald commented:
@xcompass I was hoping to give this a thumbs up just now but I found a couple little issues, including one that causes a 500 error. Should be good to go after you fix those! 

BTW I had to use a different cURL command to obtain an access code than given in your test instructions. The one that worked for me is:

```
curl --data "grant_type=authorization_code&code=CODE&client_id=CLIENT_ID&client_secret=CLIENT_SECRET" http://localhost:8000/oauth2/access_token
```

e-kolpakov commented:
@xcompass Braden is away today, so I'm going to do the reviews today; the changes since last Braden's review seem to address his points, but CI reports test failures in tests added in this PR - could you please fix them?

e-kolpakov commented:
@xcompass :+1: conditional tests are fixed and build passes

xcompass commented:
> BTW I had to use a different cURL command to obtain an access code than given in your test instructions. The one that worked for me is:

Thanks. Updated the instruction.

xcompass commented:
@e-kolpakov Updated based on Braden's comments. Tests are working now.

antoviaque commented:
@xcompass Thanks! So I think things are good now from our perspective - you've got a :+1: from @e-kolpakov and @sarina already - you might still be needing one from the platform team, and maybe @nasthagiri ? Who will be merging this PR at the end btw?

sarina commented:
@antoviaque Someone on the platform team should review and merge this, unless Nimisha feels comfortable doing it herself.

nasthagiri commented:
@xcompass I am now reviewing your PR.  Overall, it looks very good.  I have a few questions that I'll ask inline.

nasthagiri commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -545,3 +558,22 @@ def get_lti_consumer_secret(self):
+    provider_id = models.CharField(
+        max_length=255,
+        help_text=(
+            'Uniquely identify a provider. This is different from backend_name.'
```
Is this intended to be the value of `ProviderConfig.provider_id`?

xcompass replied:
Yes. it is.

nasthagiri commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -89,3 +98,122 @@ def get(self, request, username):
+        usernames_str = self.request.QUERY_PARAMS.get('usernames', None)
+        remote_ids_str = self.request.QUERY_PARAMS.get('remote_ids', None)
+
+        query_set = UserSocialAuth.objects.select_related('user').filter(provider=self.provider.backend_name)
```
If there are multiple providers using the same backend, could this result in retrieving users from other providers using the same backend?

nasthagiri replied:
For example, consider when both UniversityX and UniversityY, as providers, want to both use Google authentication for their users.  Won't UniversityX be able to see UniversityY's users since the backend_name will be the same?

xcompass replied:
@nasthagiri The provider backend_name will be different for on campus authentication. E.g. `ubcsaml`, `mitsmal`. It's defined when creating the SAML provider for each university. However It is different from your example. In your example, the actually provider is Google. When Edx enabled google, anyone can login with google. There is no UniversityX and UniversityY involvement. And Edx wouldn't want to create a API permission entry to link any oauth2 client to Google provider ID. Though for UBC, Edx may create an oauth2 client and link `ubcsaml` to that client. Then we can query all usernames and remote_id that are linked with `ubcsaml`, but not the ones with `mitsaml`. Does it make sense?

nasthagiri replied:
Yes, I understand that this API is designed to work specifically for the SAML use case.  My concern is that the API and admin interface is very general and an admin user could mistakenly create a provider that integrates with a backend that isn't designed to be campus-specific, without realizing its implications.

The `SAMLProviderConfig` has a notion of `idp_slug` and uniquely identifying itself by the provider that is using it.  However, that is not the case for `OAuth2ProviderConfig`.

Any configured OAuth backend (Google, Facebook, LinkedIn, DropBox, etc..) will be available in the django admin dropdown when configuring permissions for a Provider.  What precaution is in place that prevents an admin from making this error?

Is there a relatively easy way to disable this feature for non-SAML providers?

bradenmacdonald replied:
@nasthagiri @xcompass My understanding (and strong preference) is that this API is generic... It should support OAuth providers as well as SAML providers, and maybe even LTI providers.

For example, someone might have:
- "Google" provider (ID "oa2-google", backend_name "google-oauth2")
- "UBC" provider (ID "saml-ubc", backend_name "tpa-saml")
- "MIT" provider (ID "saml-mit", backend_name "tpa-saml")
- "Small University" provider (ID "oa2-smallu", backend_name "smallu") - a custom OAuth provider for a school that doesn't use SAML

Note the backend_name is actually the same for UBC and MIT in this example... It's the line `query_set.filter(uid__startswith=uid[:-3])` that can distinguish between different schools using the same backend. (And permissions are based on the provider ID not the backend_name)

xcompass replied:
@bradenmacdonald Thanks for the clarification. @nasthagiri sorry for the bad example, Braden's example is much better. The API is designed to be generic. Not only SAML backend can be queried but other backends as well. Since Google, Facebook, LinkedIn, DropBox don't need multi-provider support, (only one Google, etc) They don't need a slug. And their backend_name and provider ID will be different from each other even though they are all using oauth2, e.g. google-oauth2, facebook-oauth2, etc. When an admin configures the provider permissions, they will see "Google" instead of "OAuth2" for the provider. So if he/she doesn't explicitly associate a client to Google, no one will be able to query google remote IDs.

nasthagiri replied:
@bradenmacdonald @xcompass   What prevents an operator from configuring the providers as follows?

- "UniversityX" provider (ID "uni-x-google", backend_name "google-oauth2")
- "UniversityY" provider (ID "uni-y-google", backend_name "google-oauth2")

I am thumbs-up on the design for SAML-backed providers, however, I have concerns about the OAuth-backed providers.  (I don't know the details about LTI right now to determine whether it falls into the SAML or the OAuth boat.)

For SAML, I agree that it's implemented such that a single backend supports multiple providers via the use of the `idp_slug`.  And that implementation works fine for this PR.

For Google and other OAuth2 providers, however, this is not the case.  We would want to support multiple campuses/universities using, for example, Google apps, as their authentication backend for their users.  At that time, we will want to compartmentalize students/users according to their university.  Whether we do that based on the domain in their email address or some other mechanism, is yet unclear.

Since the current OAuth2 backend doesn't support multiple providers and since it's out of the scope for this PR, my recommendation is to disable it at this time.  Otherwise, supporting it has the potential of information leakage of student accounts from one provider to another.

I am CCing our devOps director, @e0d, as an FYI and for any additional input.

xcompass replied:
@nasthagiri I see your point. I've updated the dropdown to only show the SAML backends.

bradenmacdonald replied:
@nasthagiri I see your point too, however I think you are missing a few pieces of the picture. The example you gave ("UniversityX" and "UniversityY") is currently impossible because I designed the OAuth provider configuration models to be uniquely keyed to the backend. So you can only ever have one "google-oauth2" provider configured on any given Open edX instance,. This is to avoid exactly the kind of problems you are worried about.

In addition:

* Due to the way the python-social-auth `UserSocialAuth` table works, the only way to implement multiple providers with the same backend is to use a `uid` prefix. (Or creating a whole new table to JOIN with but let's not go there.) So the current solution for querying the API for multiple providers is very generic. SAML uses `idp_slug` as the `uid prefix`, but other future implementations (such as a new sharable Google OAuth2 backend) could use some other key (such as a required domain name) as the `uid prefix`. Pan's code for this was perfectly generic and flexible. I believe that changing it to explicitly mention `idp_slug`s is making it less generic and thus less useful.

* I also am not in favour of making the admin SAML-only for now. As an example, right now, a school that uses Google Apps and has their own Open edX instance could configure the google provider (with the setting `WHITELISTED_DOMAINS = ["myschool.edu"]` which rejects other domains). With this API, they would then be able to use any student's Google Apps email address to lookup each student's edX account. But if you're making this SAML-only or SAML-specific for now, you're taking away that functionality.

xcompass replied:
@nasthagiri I thought you can configure two providers both using google-oauth2. Actually, you can't. The second one will replace the first one. So the scenario in your example will never happen. You can only have one google-oauth2 backended provider.  Thanks @bradenmacdonald

nasthagiri replied:
@bradenmacdonald Thanks for the clarification on the OAuth provider.  You have confirmed that the code does, in fact, prevent an operator from configuring multiple providers to the same OAuth backend.  I was able to find the [code that retrieves only the last configured OAuth provider](https://github.com/edx/edx-platform/blob/master/common/djangoapps/third_party_auth/provider.py#L20).  

(Nit: It would be nice to add a comment there to explain the security rationale for doing this so another developer doesn't accidentally change the implementation without addressing these concerns.)

Given this, I'm now cool with the design.  :)

nasthagiri commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -89,3 +98,122 @@ def get(self, request, username):
+
+        query = Q()
+        if usernames_str:
+            usernames = usernames_str.split(',')
```
The general web standard for supplying multiple values to a URL parameter is to repeat the field in the query in order to support HTML forms:  `http://..api../users?username=alice&username=bob`
This is how [django's QueryDict](https://docs.djangoproject.com/en/1.8/ref/request-response/#querydict-objects) is meant to be used.

nasthagiri replied:
If it helps, here's what I'm doing in a different PR: https://github.com/edx/edx-platform/pull/10327/files#diff-96179cd45d4590b0a795160db721dad0R5
I've created a `MultiValueField` class that supports both comma-separated fields and multi-query parameters.

nasthagiri commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -89,3 +98,122 @@ def get(self, request, username):
+
+        # permission checking. We allow both API_KEY access and OAuth2 client credential access
+        if not (
+                ApiKeyHeaderPermission().has_permission(self.request, self) or
```
Nit: The `UserView.get` implementation for verifying the API_KEY is different from this behavior.  It would be nice to have a consistent story of how we allow and handle API_KEYs.  In particular, for `UserView.get`, it also requires that the user is a superuser.

xcompass replied:
I added superuser logic. However, I'm not sure if it is useful since it's intended usage is for server-to-server. And the logic for superuser and API_KEY in `UserView.get` is `or`. It allows the user itself, superuser or request with API_KEY to query.

nasthagiri commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -89,3 +98,122 @@ def get(self, request, username):
+    def get_queryset(self):
+        self.provider = Registry.get(self.kwargs.get('provider_id'))
+        if not self.provider:
+            raise Http404
```
The check for provider existence should come after the permission check below.  Otherwise, this can potentially leak information to unauthorized users.

nasthagiri commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -89,3 +98,122 @@ def get(self, request, username):
+            * remote_id: The Id from third party auth provider
+    """
+    authentication_classes = (
+        OAuth2Authentication,
```
When the API_KEY access path is used, how does the authentication check succeed if it requires OAuth?

nasthagiri commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -89,3 +98,122 @@ def get(self, request, username):
+
+class UserMappingView(ListAPIView):
+    """
+    Map between the third party auth account IDs (remote_id) and EdX username.
```
@sarina  Do you know the convention for using the term edX in our docstrings?  Should we be using `open edX` instead?

nasthagiri replied:
@lamagnifica @catong FYI on new API - although it's currently at v0.

nasthagiri commented on common/djangoapps/third_party_auth/admin.py:
```diff
@@ -2,12 +2,21 @@
+    SAMLConfiguration,
+    SAMLProviderData,
+    LTIProviderConfig,
+    ApiPermissions
```
Can we rename this to `ProviderApiPermissions` so it is it clearly scoped in its name?

nasthagiri commented on common/djangoapps/third_party_auth/api/permissions.py:
```diff
@@ -0,0 +1,29 @@
+        Check if the OAuth client associated with auth token in current request has permission to access
+        the information for provider
+        """
+        provider_id = request.parser_context['kwargs'].get('provider_id', None)
```
Nit: Consider passing the `provider_id` in the `__init__` method instead of assuming it's a certain field in the request object.

nasthagiri commented on common/djangoapps/third_party_auth/api/permissions.py:
```diff
@@ -0,0 +1,29 @@
+from third_party_auth.models import ApiPermissions
+
+
+class ThirdPartyAuthApiPermission(permissions.BasePermission):
```
Can we rename this to `ProviderApiPermission` since it's scoped to having access to a provider?

nasthagiri commented on common/djangoapps/third_party_auth/api/serializers.py:
```diff
@@ -0,0 +1,23 @@
+
+    def get_remote_id(self, social_user):
+        """ Gets remote id from social user based on provider """
+        assert self.provider is not None
```
Is this `assert` necessary?

nasthagiri commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -89,3 +98,122 @@ def get(self, request, username):
+
+        # build our query filters
+        # When using multi-IdP backend, we only retrieve the ones that are for current IdP
+        uid = self.provider.get_social_auth_uid('uid')
```
Why have a method that prepends the `idp_slug` to a dummy string, which you then slice out 2 lines below?  Since `get_social_auth_uid` is a new method created in this PR, why not have it return exactly what's needed, which is the provider's idp_slug?

xcompass replied:
This is basically testing if the there is any idp_slug for this provider. There are two types of provider, one like google, only one provider, one backend. Another is SAML, one SAML backend but multiple providers, e.g. UBC SAML, MIT SAML. Because social auth table doesn't have a column to support multi-provider backend. The implementation for SAML is prepending the slug in the remote_id column to distinguish between different providers with the same backend. The following two lines are adding a filter if the current query is for a multi-provider backend.

nasthagiri replied:
Yes, I understand that the SAML implementation uses an `idp_slug`.  My comment is about making that explicit in the call you are making.  

For example, instead of creating a method called `get_social_auth_uid`, can you create a method called `get_idp_slug`?  This way, it will be exactly clear that you are getting the provider's `idp_slug`.  This would be better than creating an uid for a dummy remote_id and then immediately parsing it.

nasthagiri commented on common/djangoapps/third_party_auth/tests/testutil.py:
```diff
@@ -114,6 +117,18 @@ def verify_user_email(cls, email):
+        """ Configure the client and provider_id pair. This will give the access to a client for that provider. """
+        obj = ApiPermissions(client=client, provider_id=provider_id)
+        obj.save()
+        return obj
```
Nit: To create and save an object in one step use the [create method](https://docs.djangoproject.com/en/dev/ref/models/querysets/#django.db.models.query.QuerySet.create).  So the 3 lines above would become:

`return ApiPermissions.objects.create(client=client, provider_id=provider_id)`

nasthagiri commented on common/djangoapps/third_party_auth/api/tests/test_permissions.py:
```diff
@@ -0,0 +1,46 @@
+        super(ThirdPartyAuthApiPermissionTest, self).setUp()
+
+        self.enable_saml()
+        self.configure_saml_provider(name='TestShib', enabled=True, idp_slug=IDP_SLUG_TESTSHIB)
```
Is enabling and configuring `saml` even required for these tests?  The test creates and validates the `ApiPermissions` objects that are created, but doesn't integrate with any actual providers.
If my read is correct, then please delete the saml configuration lines in the setup.

nasthagiri commented on common/djangoapps/third_party_auth/api/tests/test_permissions.py:
```diff
@@ -0,0 +1,46 @@
+        (1, 'invalid-provider-id', False),
+        (999, PROVIDER_ID_TESTSHIB, False),
+        (999, 'invalid-provider-id', False),
+        (1, None, False),
```
Thanks for having negative test cases for invalid providers.
Can you also add a negative test case for an existing valid provider but that is not accessible by the requesting client?

xcompass replied:
Added

nasthagiri replied:
Thanks.

nasthagiri commented on common/djangoapps/third_party_auth/api/tests/test_views.py:
```diff
@@ -130,3 +135,83 @@ def test_list_connected_providers__withapi_key(self, api_key, target_user, expec
             self.assertIn("active", response.data)
             self.assertItemsEqual(response.data["active"], self.expected_active(target_user))
+
+    @ddt.data(
```
Please create a new Test class for testing the `UserMappingView`.  You can rename the existing one from `ThirdPartyAuthAPITests` to something like `TestUserView`.  You can also move any common `setUp`s and helpers into a common Test base or mixin class.

nasthagiri commented on common/djangoapps/third_party_auth/api/tests/test_views.py:
```diff
@@ -1,26 +1,31 @@
+
+def get_mapping_data_by_usernames(usernames):
+    """ Generate mapping data used in response """
+    return [{'username': username, 'remote_id': username} for username in usernames]
```
Can you update the test to verify that the `remote_id` is actually returning the remote id and not just the open edX username?  You can do so by updating the call to `UserSocialAuth.objects.create` on L71 and passing in a modified username for the user's remote ID.

nasthagiri commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -89,3 +98,122 @@ def get(self, request, username):
+            * username: The edx username
+
+            * remote_id: The Id from third party auth provider
+    """
```
Please add a comment somewhere that this is intended to be a server-to-server endpoint.

xcompass replied:
Added at the top of the docstring

nasthagiri commented on common/djangoapps/third_party_auth/api/tests/test_views.py:
```diff
@@ -130,3 +135,83 @@ def test_list_connected_providers__withapi_key(self, api_key, target_user, expec
+        response = self.client.get(url, HTTP_X_EDX_API_KEY=VALID_API_KEY)
+        self.assertEqual(response.status_code, expect_result)
+        if expect_result == 200:
+            self._verify_valid_response(response, ['results'], expect_data)
```
Why are 'count' and 'num_pages' not included here?
Consider creating a helper function for the last 3 lines since it is repeated in several user-mapping related tests.

nasthagiri commented on common/djangoapps/third_party_auth/api/tests/test_views.py:
```diff
@@ -130,3 +135,83 @@ def test_list_connected_providers__withapi_key(self, api_key, target_user, expec
+    @ddt.unpack
+    def test_user_mappings_with_query_params(self, query_params, expect_result, expect_data):
+        base_url = reverse(
+            'third_party_auth_user_mapping_api', kwargs={'provider_id': PROVIDER_ID_TESTSHIB})
```
Nit: format close paren to next line.

nasthagiri commented on common/djangoapps/third_party_auth/api/tests/test_views.py:
```diff
@@ -130,3 +135,83 @@ def test_list_connected_providers__withapi_key(self, api_key, target_user, expec
+        for attr in ['usernames', 'remote_ids']:
+            if attr in query_params:
+                params.append('{}={}'.format(attr, ','.join(query_params[attr])))
+        url = "{}?{}".format(base_url, '&'.join(params))
```
Nit: consider using a `QueryDict` object instead and calling `setlist`.  Here's an example: https://github.com/edx/edx-platform/blob/062e979e0ed14161d0f882091a06e032ca76bb3e/lms/djangoapps/discussion_api/tests/test_forms.py#L105

xcompass replied:
I've added a separate test using QueryDict to test multi-valued fields.

nasthagiri commented on common/djangoapps/third_party_auth/api/tests/test_views.py:
```diff
@@ -130,3 +135,83 @@ def test_list_connected_providers__withapi_key(self, api_key, target_user, expec
+        params = []
+        for attr in ['usernames', 'remote_ids']:
+            if attr in query_params:
+                params.append('{}={}'.format(attr, ','.join(query_params[attr])))
```
Once again, to support HTML Form elements, notably <select multiple>, we'll need to support dict entries for the same key.

nasthagiri commented on common/djangoapps/third_party_auth/api/tests/test_views.py:
```diff
@@ -130,3 +135,83 @@ def test_list_connected_providers__withapi_key(self, api_key, target_user, expec
         if expect_result == 200:
             self.assertIn("active", response.data)
             self.assertItemsEqual(response.data["active"], self.expected_active(target_user))
+
```
Can you add some (non-mocked) test cases for the OAuth2 authentication path?

xcompass replied:
Added

nasthagiri commented:
@xcompass I have completed my first pass.

xcompass commented:
@nasthagiri updated based on your comments.

nasthagiri commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -89,3 +98,145 @@ def get(self, request, username):
+        usernames_str = self.request.QUERY_PARAMS.get('usernames', None)
+        remote_ids_str = self.request.QUERY_PARAMS.get('remote_ids', None)
+        username_list = self.request.QUERY_PARAMS.getlist('username', None)
+        remote_ids_list = self.request.QUERY_PARAMS.getlist('remote_id', None)
```
This presents a confusing API if there are multiple ways to provide a list of usernames and a list of remote_ids. 
The `QueryDict` implementation already supports both comma-separated and multi-value fields.  Can we just support the last 2 queries above?

xcompass replied:
Could you point me to where the `QueryDict` supporting comma-separated fields? I can't find any mentioned comma in the doc. Also the field name is slightly different (username vs usernames). Should I use the same field name (singular form)? (/user?username=user1,user2)

nasthagiri replied:
@xcompass Yes, please use the same field name: singular form.

I see the problem you are running into.  In a separate (but not yet merged) PR, I have created a common [MultiValueField](https://github.com/edx/edx-platform/pull/9292/files#diff-96179cd45d4590b0a795160db721dad0R5) class that serves exactly this purpose.  It supports both ways of passing in multiple values for a filed.

That class is not yet merged and assumes one is using django Forms for parsing the query params.  In the interim, though, you can use the same logic.  Essentially:
1. join all the fields into one comma-separated string, and then
2. parse the comma-separated string into multiple values in a set/list.

nasthagiri commented on common/djangoapps/third_party_auth/api/tests/test_views.py:
```diff
@@ -130,3 +144,129 @@ def test_list_connected_providers__withapi_key(self, api_key, target_user, expec
+        user = UserFactory.create(username='api_user')
+        client = Client.objects.create(name='oauth2_client', client_type=CONFIDENTIAL)
+        token = AccessToken.objects.create(user=user, client=client)
+        ProviderApiPermissions.objects.create(client=client, provider_id=provider_id)
```
Can you add a negative test case here for OAuth2 to make sure the `ProviderApiPermissions` is actually enforcing when an unauthorized client accesses a provider?
Suggestion: You can add another `ddt` parameter for using an invalid access token.

nasthagiri commented:
@xcompass Thanks for updating the PR based on the last round of review.  It looks great!  I'm now waiting for the following items before giving my thumbs:
1. Updating the multi-value field code per suggestion (and only supporting the singular form).
2. Negative test case for OAuth2.
3. Squashing and final cleanup of the commits.

xcompass commented:
@nasthagiri All done. Let me know if there is anything else. Looking forward to get this merge :) Thanks for all your help.

nasthagiri commented on common/djangoapps/third_party_auth/api/tests/test_views.py:
```diff
@@ -130,3 +144,133 @@ def test_list_connected_providers__withapi_key(self, api_key, target_user, expec
+        ({'username': [ALICE_USERNAME], 'remote_id': ['remote_' + STAFF_USERNAME]}, 200,
+         get_mapping_data_by_usernames([ALICE_USERNAME, STAFF_USERNAME])),
+    )
+    @ddt.unpack
```
Are there any tests for testing invalid usernames and invalid remote_ids?

xcompass replied:
Added

nasthagiri commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -89,3 +97,136 @@ def get(self, request, username):
+        It is suggested to query no more than 50 usernames or remote_ids in each request to stay within above
+        limitation
+
+        Each page in the list can contain up to 50 mappings.
```
Isn't this governed by the `page_size`?  Or is 50 a hard limit?

xcompass replied:
It is actually by page_size. I've updated the docstring.

On Tue, Oct 27, 2015 at 2:56 PM Nimisha Asthagiri <notifications@github.com>
wrote:

> In common/djangoapps/third_party_auth/api/views.py
> <https://github.com/edx/edx-platform/pull/9842#discussion_r43193347>:
>
> > +class UserMappingView(ListAPIView):
> > +    """
> > +    Map between the third party auth account IDs (remote_id) and EdX username.
> > +
> > +    This API is intended to be a server-to-server endpoint. An on-campus middleware or system should consume this.
> > +
> > +    **Use Case**
> > +
> > +        Get a paginated list of mappings between edX users and remote user IDs for all users currently
> > +        linked to the given backend.
> > +
> > +        The list can be filtered by edx username or third party ids. The filter is limited by the max length of URL.
> > +        It is suggested to query no more than 50 usernames or remote_ids in each request to stay within above
> > +        limitation
> > +
> > +        Each page in the list can contain up to 50 mappings.
>
> Isn't this governed by the page_size? Or is 50 a hard limit?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/9842/files#r43193347>.
>

nasthagiri commented:
:+1:

xcompass commented:
@sarina Nimisha finished the review. I have squashed the commits and rebased on current master.  Let me know if there is anything else need to be done.

nasthagiri commented:
@xcompass Looks good.  Thanks for all your work on this PR.  It's a great contribution to open edX and our partners.  Is there anything left on your end?  If not, I'll go ahead and merge.

xcompass commented:
@nasthagiri I don't think so. Thanks. Glad to make the contribution and I learnt a lot :)

nasthagiri commented:
FYI - @clintonb has added support for the `client_credentials` grant type in our django oauth2 provider.  See https://github.com/edx/django-oauth2-provider/pull/25.  Given this new support, we should now be able to update this feature to use client_credentials as well.

xcompass commented:
@nasthagiri Great! I'll take a look at it and update the API in a new PR

xcompass commented:
@nasthagiri It looks I don't need to update anything :)

Is this in production already? (Edge and edx.org) I'll update the PR message to reflect the change.

nasthagiri commented:
@xcompass It will be in production later today (hopefully by noon).  You shouldn't need to update any code.  However, you can (1) eliminate the use of the `EDX_API_KEY` and (2) update your OAuth call to use the `client_credentials` grant (assuming your OAuth client is already configured as a "Confidential" client).

xcompass commented:
@nasthagiri Thanks! That's what I thought.
